source/extensions/filters/http/decompressor/decompressor_filter.cc

```diff
@@ -71,7 +71,27 @@ Http::FilterHeadersStatus DecompressorFilter::decodeHeaders(Http::RequestHeaderM
   //      the upstream that this hop is able to decompress responses via the Accept-Encoding header.
   if (config_->responseDirectionConfig().decompressionEnabled() &&
       config_->requestDirectionConfig().advertiseAcceptEncoding()) {
-    headers.appendInline(accept_encoding_handle.handle(), config_->contentEncoding(), ",");
+    // Append the content encoding only if it isn't already present in the
+    // accept_encoding header. If it is present with a q-value ("gzip;q=0.3"),
+    // remove the q-value to indicate that the content encoding setting that we
+    // add has max priority (i.e. q-value 1.0).
+    std::vector<absl::string_view> newContentEncodings;
+    std::vector<absl::string_view> contentEncodings = StringUtil::splitToken(
+        headers.getInlineValue(accept_encoding_handle.handle()), ",", true, true);
+    for (absl::string_view contentEncoding : contentEncodings) {
+      absl::string_view strippedEncoding =
+          StringUtil::trim(StringUtil::cropRight(contentEncoding, ";"));
```

Comment:
This looks good, but can you factor out this logic into a helper function that you can
 * put in source/common/http/header_utility.cc along with similar functions
 * thoroughly unit-test in test/common/http/header_utility_test.cc

That way we'll centralize the code that knows a lot about http headers value syntax.

I actually want someone involved with caching to do a first pass on this as they also (likely) do a lot of http header analysis and it would be good to avoid duplicating effort, especially as these kinds of things tend to be easy to get wrong :)

Reply:
There is some similar code in CacheHeaderUtils alongside the CacheFilter.

https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/cache/cache_headers_utils.cc#L218

I agree that a common header processing util would make sense here, and perhaps things like parsing comman delimited headers should go along with it.

Reply:
Also as a data plane change we should probably runtime guard this.
/wait

Reply:
Thanks for the comments! I'll try first moving the CacheHeaderUtils parsing function to HeaderUtility and adding tests to it. I don't plan to implement the `;`-separated value parsing because I don't think this use case really needs to have it.